bitcoin/src/internal_macros.rs

```diff
@@ -211,3 +211,108 @@ macro_rules! impl_asref_push_bytes {
     };
 }
 pub(crate) use impl_asref_push_bytes;
+
+/// Defines an trait `$trait_name` and implements it for `ty`, used to define extension traits.
+macro_rules! define_extension_trait {
+    // With self, no generics.
+    ($(#[$($trait_doc_attrs:tt)*])* $trait_vis:vis trait $trait_name:ident impl for $ty:ident {
+        $(
+            $(#[$($fn_doc_attrs:tt)*])*
```

Comment:
Note that this captures all attributes, not just doc ones, but that's actually good since we can use `#[inline]`. I'd just avoid writing `doc_` in the variable name to avoid confusion.

Reply:
Oh that reminds me, I didn't comment - I removed the `#[inline]` because I got build warnings. I suppose we could split the attributes and put the inline in the trait implementation and not in the trait declaration, I kind of can't be bothered though. Do you think its worth the trouble?

But yes to the remove `_doc` from the name.

Reply:
Oh, right, but we need to put doc on trait and don't need to on the impl.

I think many of the methods will benefit from inlining but I think it's OK to postpone it.

Reply:
This is resolved, right? (Attributes are currently on the trait method declarations only. Will need further modification to support inline.)